src/video/gdk/SDL_gdktextinput.cpp

```diff
@@ -0,0 +1,311 @@
+     appreciated but is not required.
+  2. Altered source versions must be plainly marked as such, and must not be
+     misrepresented as being the original software.
+  3. This notice may not be removed or altered from any source distribution.
+*/
+/*
+  Screen keyboard and text input backend
+  for GDK platforms.
+*/
+#include "SDL_internal.h"
+#include "SDL_gdktextinput.h"
+
+#ifdef SDL_GDK_TEXTINPUT
+
+/* GDK headers are weird here */
+#ifndef WIN32_LEAN_AND_MEAN
+#define WIN32_LEAN_AND_MEAN
+#endif
+#include <Windows.h>
+#include <XGameUI.h>
+#include <XUser.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#include "../../events/SDL_keyboard_c.h"
+#include "../windows/SDL_windowsvideo.h"
+
+/* TODO: Have a separate task queue for text input perhaps? */
+static XTaskQueueHandle g_TextTaskQueue = NULL;
+/* Global because there can be only one text entry shown at once. */
+static XAsyncBlock *g_TextBlock = NULL;
+
+/* Creation parameters */
+static SDL_bool g_DidRegisterHints = SDL_FALSE;
+static char *g_TitleText = NULL;
+static char *g_DescriptionText = NULL;
+static char *g_DefaultText = NULL;
+static const Sint32 g_DefaultTextInputScope = (Sint32)XGameUiTextEntryInputScope::Default;
+static Sint32 g_TextInputScope = g_DefaultTextInputScope;
+static const Sint32 g_DefaultMaxTextLength = 1024; /* as per doc: maximum allowed amount on consoles */
+static Sint32 g_MaxTextLength = g_DefaultMaxTextLength;
+
+static void SDLCALL GDK_InternalHintCallback(
+    void *userdata,
+    const char *name,
+    const char *oldValue,
+    const char *newValue)
+{
+    if (userdata == NULL) {
+        return;
+    }
+
+    /* oldValue is ignored because we store it ourselves. */
+    /* name is ignored because we deduce it from userdata */
+
+    if (userdata == &g_TextInputScope || userdata == &g_MaxTextLength) {
+        /* int32 hint */
+        Sint32 intValue = (newValue == NULL || newValue[0] == '\0') ? 0 : SDL_atoi(newValue);
+        if (userdata == &g_MaxTextLength && intValue <= 0) {
+            intValue = g_DefaultMaxTextLength;
+        } else if (userdata == &g_TextInputScope && intValue < 0) {
+            intValue = g_DefaultTextInputScope;
+        }
+
+        *(Sint32 *)userdata = intValue;
+    } else {
+        /* string hint */
+        if (newValue == NULL || newValue[0] == '\0') {
+            /* treat empty or NULL strings as just NULL for this impl */
+            SDL_free(*(char **)userdata);
+            *(char **)userdata = NULL;
+        } else {
+            char *newString = SDL_strdup(newValue);
+            if (newString == NULL) {
+                /* couldn't strdup, oh well */
+                SDL_OutOfMemory();
+            } else {
+                /* free previous value and write the new one */
+                SDL_free(*(char **)userdata);
+                *(char **)userdata = newString;
+            }
+        }
+    }
+}
+
+static void GDK_InternalEnsureHints(void)
```

Comment:
Is there a place we can call this from the GDK video initialization instead of sprinkling calls to this all over?

Reply:
Maybe like this (*force push*)? If thread-safety isn't a huge concern this should be fine. (afaik SDL's video subsystem is not thread-safe anyway)